core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -147,6 +147,30 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
         }
     )(use)
 
+  /**
+   * Returns an effect that, if evaluated, will return the cached result of
+   * this effect. Cached results will expire after `timeToLive` duration.
+   */
+  final def cached(timeToLive: Duration): ZIO[R with Clock, Nothing, IO[E, A]] = {
+
+    def get(cache: RefM[Option[Promise[E, A]]]): ZIO[R with Clock, E, A] =
+      cache.update {
+        case Some(p) =>
+          ZIO.succeed(Some(p))
+        case None =>
+          for {
+            p <- Promise.make[E, A]
+            _ <- self.to(p)
+            _ <- p.await.delay(timeToLive).flatMap(_ => cache.set(None)).fork
+          } yield Some(p)
+      }.flatMap(_.get.await)
```

Comment:
Looks great! There are only two ways to improve this:

 - Make the ref a try state (empty, pending, or full), so two fibers don't try to fill it at the same time
 - Deal with interruption, i.e. ensure interruption of a `get` does not prevent the cache from _ever_ being filled

The first would necessitate the second, I think.

In any case, looks good to me, 👍  to ship!

Reply:
I updated it to deal with interruption by using `uninterruptibleMask` so only the waiting on the promise is interruptible. Regarding your first point, I thought I had addressed this by using `RefM`. My thinking was that if two fibers called `get` at the same time the first would create the promise, fork the fiber to fill it, and take a reference to that promise, and then the second would wait for the promise creation / forking and take a copy of the same promise. So the empty / pending / full states would correspond to none, some pending promise, and some completed promise. Does this work or am I messing something up here?

Reply:
Ah, yes, you did. I forgot you were using `RefM`. Effects will be synchronized appropriately.